File: python/gradbench/pytorch/lstm.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

# https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/python/modules/PyTorch/PyTorchLSTM.py

"""
Changes Made:
- Added two functions to create a PyTorchHand object and call calculate_objective and calculate_jacobian
"""

import time

import numpy as np
import torch

from gradbench.adbench.itest import ITest
from gradbench.adbench.lstm_data import LSTMInput, LSTMOutput
from gradbench.pytorch.lstm_objective import lstm_objective
from gradbench.pytorch.utils import to_torch_tensors, torch_jacobian
from gradbench.wrap_module import wrap


class PyTorchLSTM(ITest):
    """Test class for LSTM diferentiation by PyTorch."""

    def prepare(self, input):
        """Prepares calculating. This function must be run before
        any others."""

        self.inputs = to_torch_tensors(
            (input.main_params, input.extra_params), grad_req=True
        )

        self.params = to_torch_tensors((input.state, input.sequence))
        self.gradient = torch.empty(0)
        self.objective = torch.zeros(1)

    def output(self):
        """Returns calculation result."""

        return LSTMOutput(self.objective.item(), self.gradient.numpy())

    def calculate_objective(self, times):
        """Calculates objective function many times."""

        for i in range(times):
            self.objective = lstm_objective(*self.inputs, *self.params)

    def calculate_jacobian(self, times):
        """Calculates objective function jacobian many times."""

        for i in range(times):
            self.objective, self.gradient = torch_jacobian(
                lstm_objective, self.inputs, self.params
            )


def prepare_input(input):
    py = PyTorchLSTM()
    py.prepare(LSTMInput.from_dict(input))
    return py, input["runs"]


def evaluate_times(times):
    return [{"name": "evaluate", "nanoseconds": time} for time in times]


def unwrap_objective(output):
    py, times = output
    return float(py.objective.detach().flatten().numpy()[0]), evaluate_times(output[1])


def unwrap_jacobian(output):
    py, times = output
    return py.gradient.tolist(), evaluate_times(times)


@wrap(prepare_input, unwrap_objective)
def objective(input):
    py, runs = input
    times = runs * [None]
    for i in range(runs):
        start = time.perf_counter_ns()
        py.calculate_objective(runs)
        end = time.perf_counter_ns()
        times[i] = end - start
    return py, times


@wrap(prepare_input, unwrap_jacobian)
def jacobian(input):
    py, runs = input
    times = runs * [None]
    for i in range(runs):
        start = time.perf_counter_ns()
        py.calculate_jacobian(runs)
        end = time.perf_counter_ns()
        times[i] = end - start
    return py, times

File: tools/tapenade/lstm.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// Largely derived from https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/cpp/modules/tapenade/TapenadeLSTM.cpp

#include <algorithm>

#include "gradbench/main.hpp"
#include "gradbench/evals/lstm.hpp"

#include "lstm/lstm.h"
#include "lstm/lstm_b.h"

class Jacobian : public Function<lstm::Input, lstm::JacOutput> {
  std::vector<double> _state;

public:
  Jacobian(lstm::Input& input) : Function(input) {}

  void compute(lstm::JacOutput& output) {
    output.resize(8 * _input.l * _input.b + 3 * _input.b);
    double* main_params_gradient_part = output.data();
    double* extra_params_gradient_part = output.data() + _input.main_params.size();

    double loss = 0.0;      // stores fictive output
    // (Tapenade doesn't calculate an original function in reverse mode)

    double lossb = 1.0;     // stores dY
    // (equals to 1.0 for gradient calculation)

    _state = _input.state;
    lstm_objective_b(
                     _input.l,
                     _input.c,
                     _input.b,
                     _input.main_params.data(),
                     main_params_gradient_part,
                     _input.extra_params.data(),
                     extra_params_gradient_part,
                     _state.data(),
                     _input.sequence.data(),
                     &loss,
                     &lossb
                     );

  }
};

int main(int argc, char* argv[]) {
  return generic_main(argc, argv, {
      {"objective", function_main<lstm::Objective>},
      {"jacobian", function_main<Jacobian>}
    });;
}
